### PR TITLE
Remove cookies

### LINKED ISSUE #505: Remove cookies

### PR DESCRIPTION
**Pull request by komninoschat**
Resolves #505.

### CODE REVIEW

File: src/middlewares/auth.ts [hunk 1]
```diff
@@ -34,32 +26,13 @@ export function authMiddleware(req: RequestExtended, res: Response, next: NextFu
...
-
-  if ('permission_variables' in cookiesInUse) {
-    try {
-      req.permission_variables = getPermissionVariablesFromCookie(req)
-    } catch (err) {
-      return res.boom.unauthorized(err.message)
-    }
+  if ('permission_variables' in req.query) {
+    req.permission_variables = JSON.parse(req.query.permission_variables as string)
+    delete req.query.permission_variables
```
Comment (elitan) [inline src/middlewares/auth.ts:35]: Why do we do this? Looks like a security issue if the permission variables can be set using a query variable.

File: src/routes/auth/activate.ts [hunk 2]
```diff
@@ -49,7 +49,10 @@ async function activateUser({ query }: Request, res: Response): Promise<unknown>
     return res.redirect(302, APPLICATION.REDIRECT_URL_SUCCESS)
   }
 
-  res.status(200).send('Your account has been activated. You can close this window and login')
+  if(APPLICATION.REDIRECT_URL_SUCCESS) {
+    res.redirect(APPLICATION.REDIRECT_URL_SUCCESS.replace('JWT_TOKEN', ticket))
```
Comment (elitan) [inline src/routes/auth/activate.ts:53]: What's the use case here? I feel like `JWT_TOKEN` and `ticket` should not be mixed like this?

File: src/routes/auth/providers/utils.ts [hunk 3]
```diff
@@ -119,14 +118,15 @@ const providerCallback = async (req: RequestExtended, res: Response): Promise<vo
...
   try {
-    refresh_token = await setRefreshToken(res, account.id, true)
+    [refresh_token, permission_variables] = await setRefreshToken(account.id)
   } catch (e) {
     res.redirect(PROVIDERS.REDIRECT_FAILURE)
   }
 
   // redirect back user to app url
-  res.redirect(`${PROVIDERS.REDIRECT_SUCCESS}?refresh_token=${refresh_token}`)
+  res.redirect(`${PROVIDERS.REDIRECT_SUCCESS}?refresh_token=${refresh_token}&permission_variables=${permission_variables}`)
```
Comment (elitan) [inline src/routes/auth/providers/utils.ts:129]: Why do we send `permission_variables` here?

File: src/shared/helpers.ts [hunk 4]
```diff
@@ -118,8 +118,33 @@ export const rotateTicket = async (ticket: string): Promise<string> => {
   return new_ticket
 }
 
-export const getPermissionVariablesFromCookie = (req: RequestExtended): PermissionVariables => {
-  const { permission_variables } = COOKIES.SECRET ? req.signedCookies : req.cookies
-  if (!permission_variables) throw new Error('No permission variables')
-  return JSON.parse(permission_variables)
+export function newRefreshExpiry(): number {
+  const now = new Date()
+  const days = JWT.REFRESH_EXPIRES_IN / 1440
```
Comment (elitan) [inline src/shared/helpers.ts:123]: Could we add a comment about the `1440` here to increase the code readability?

ex: // converting from minutes to days

File: src/shared/types.ts [hunk 5]
```diff
@@ -48,6 +48,7 @@ export interface Session {
   jwt_token: string | null;
   jwt_expires_in: number | null;
   refresh_token?: string
+  permission_variables?: string
```
Comment (elitan) [inline src/shared/types.ts:51]: Why is this added here?

**Review by elitan — CHANGES_REQUESTED** [review]
See previous comments.

File: src/routes/auth/magic-link.ts [hunk 6]
```diff
@@ -70,8 +69,7 @@ async function magicLink({ query }: Request, res: Response): Promise<unknown> {
     email: account.email,
     avatar_url: account.user.avatar_url
   }
-  const session: Session = { jwt_token, jwt_expires_in, user }
-  if (!useCookie) session.refresh_token = refresh_token
+  const session: Session = { jwt_token, jwt_expires_in, user, refresh_token, permission_variables }
```
Comment (elitan) [inline src/routes/auth/magic-link.ts:72]: I don't think we should return `permission_variables` here (or any other place) to the client.

File: src/routes/auth/magic-link.ts [hunk 7]
```diff
@@ -84,7 +82,10 @@ async function magicLink({ query }: Request, res: Response): Promise<unknown> {
       return res.redirect(`${APPLICATION.REDIRECT_URL_SUCCESS}?refresh_token=${refresh_token}`)
     }
 
-    res.status(200).send('Your account has been activated. You can close this window and login')
+    if(APPLICATION.REDIRECT_URL_SUCCESS) {
+      res.redirect(APPLICATION.REDIRECT_URL_SUCCESS.replace('JWT_TOKEN', token))
```
Comment (elitan) [inline src/routes/auth/magic-link.ts:86]: What is `.replace('JWT_TOKEN', token)` used for?

**Review by elitan — DISMISSED** [review]

File: src/routes/auth/activate.ts (hunk 2 above)
Comment (elitan) [reply src/routes/auth/activate.ts:53]: Let's fix in a separate issue

File: src/routes/auth/magic-link.ts (hunk 7 above)
Comment (elitan) [reply src/routes/auth/magic-link.ts:85]: Will be fixed in a separate PR.

**Comment by nikolowry** [comment]
@komninoschat @elitan will there be updated guidance on how to store the `refresh_token` in the client-side?

Per https://hasura.io/blog/best-practices-of-using-jwt-with-graphql/#refresh_token_persistance, the point of the `HttpOnly` cookie was to avoid storing it in `window.localStorage`. Is persisting the `refresh_token` somehow different than persisting the JWT token?

**Comment by elitan** [comment]
We recommend using `localStorage`: https://github.com/nhost/nhost-js-sdk/blob/master/src/NhostClient.ts#L31-L33

**Comment by nikolowry** [comment]
@elitan the best language I could find to express my concerns can be found from [Auth0's blog](https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/#Security-considerations):
> Security considerations
Refresh tokens are long-lived. This means when a client gets a refresh token from a server, this token must be stored securely to keep it from being used by potential attackers. If a refresh token is leaked, it may be used to obtain new access tokens (and access protected resources) until it is either blacklisted or it expires (which may take a long time). Refresh tokens must be issued to a single authenticated client to prevent use of leaked tokens by other parties. Access tokens must be kept secret, but as you may imagine, security considerations are less strict due to their shorter life.

Storing the refresh token in `localStorage` seems to be slightly better practice than storing the whole JWT token. It seems like security is taking a step back with this approach, or am I not fully understanding something?

**Comment by elitan** [comment]
From my perspective, storing the `refresh_token` in `localStorage` is secure enough.

But you can store the refresh token any way you want by specifying a `custom` storage strategy and provide the functions `setItem`, `getItem` and `removeItem` when you initiate the nhost js sdk:

https://github.com/nhost/nhost-js-sdk/blob/master/src/types.ts#L26-L33

**Pull request merged**